File: Masters/Semester1/NUM1/Homework/HW5/Code/p1.c
#include "../../utils/linalg/eigenvalues.h"
#include "../../utils/linalg/linear_system.h"
#include "../../utils/linalg/matrix.h"
#include "../../utils/linalg/matrix_factorization.h"
#include "../../utils/linalg/matrix_io.h"
#include "../../utils/linalg/vector.h"
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char *argv[]) {
  if (argc != 3) {
    fprintf(stderr,
            "Invalid argument count. \nUsage: %s <matrix-input> <vector-input>",
            argv[0]);
    exit(1);
  }
  int rows, cols;

  Matrix *A = load_matrix_from_txt(argv[1], &rows, &cols);
  if (!A) {
    exit(1);
  }

  Vector *b = load_vector_from_txt(argv[2]);
  if (!b) {
    matrix_free(A);
    exit(1);
  }
  double lambda = iterative_power(A, b, 1e-6, 100);
  // Using A = LU, we have LUx = b, so we solve Ly = b and then Ux = y.
  printf("%lf\n", lambda);
  matrix_free(A);
  free_vector(b);
  return 0;
}

File: Masters/Semester1/NUM1/Homework/HW7/Code/p1.py
import numpy as np
import sys

def solve_with_qr(filename_A, filename_b):
    A = np.loadtxt(filename_A)
    b = np.loadtxt(filename_b)
    x = np.linalg.solve(A, b)
    print(x)

def main():
    if (len(sys.argv) != 3):
        print("Invalid argument count.\n")
        sys.exit(1)
    fA = sys.argv[1]
    fb = sys.argv[2]
    solve_with_qr(fA, fb)
    

if __name__ == "__main__":
    main()

